File: backend/app/services/blockchain.py
"""
Ocean Sentinel - Blockchain Data Integrity Service
Starton platform integration for immutable data logging
"""

import asyncio
import aiohttp
import logging
import hashlib
import json
from datetime import datetime
from typing import Dict, List, Optional, Any
from uuid import uuid4

from app.config import settings, API_CONFIG
from app.utils.database import create_supabase_client

logger = logging.getLogger(__name__)

class BlockchainService:
    """Blockchain data integrity service using Starton platform"""
    
    def __init__(self):
        self.supabase = create_supabase_client()
        self.starton_api_key = settings.starton_api_key
        self.contract_address = settings.contract_address
        self.network = settings.polygon_network
        self.base_url = API_CONFIG['starton']['base_url']
        
        # Headers for Starton API
        self.headers = {
            'x-api-key': self.starton_api_key,
            'Content-Type': 'application/json'
        }
    
    async def log_environmental_data(
        self, 
        data_hash: str, 
        timestamp: str, 
        source: str
    ) -> Optional[str]:
        """
        Log environmental data hash to blockchain
        Args:
            data_hash: SHA-256 hash of environmental data
            timestamp: ISO timestamp of data collection
            source: Data source identifier
        Returns:
            Transaction hash if successful
        """
        try:
            logger.info(f"📝 Logging environmental data to blockchain: {source}")
            
            # Prepare transaction data
            function_name = "logEnvironmentalData"
            params = [data_hash, timestamp, source]
            
            # Execute smart contract function
            tx_hash = await self._execute_contract_function(function_name, params)
            
            if tx_hash:
                # Store blockchain transaction record
                await self._store_blockchain_record(
                    tx_hash, data_hash, timestamp, source, 'environmental_data'
                )
                
                logger.info(f"✅ Environmental data logged to blockchain: {tx_hash}")
                return tx_hash
            
            return None
            
        except Exception as e:
            logger.error(f"❌ Failed to log environmental data to blockchain: {e}")
            return None
    
    async def log_threat_data(self, threat_data: Dict, threat_id: str) -> Optional[str]:
        """
        Log threat detection to blockchain
        Args:
            threat_data: Threat detection data
            threat_id: Unique threat identifier
        Returns:
            Transaction hash if successful
        """
        try:
            logger.info(f"🚨 Logging threat data to blockchain: {threat_data.get('type')}")
            
            # Create hash of threat data
            threat_hash = self._create_data_hash(threat_data)
            
            # Prepare transaction data
            function_name = "logThreatData"
            params = [
                threat_id,
                threat_hash,
                threat_data.get('type', 'unknown'),
                threat_data.get('severity', 1),
                int(threat_data.get('confidence', 0) * 100),  # Convert to percentage
                datetime.utcnow().isoformat()
            ]
            
            # Execute smart contract function
            tx_hash = await self._execute_contract_function(function_name, params)
            
            if tx_hash:
                # Store blockchain transaction record
                await self._store_blockchain_record(
                    tx_hash, threat_hash, datetime.utcnow().isoformat(), 
                    'ai_detection', 'threat_data', threat_id
                )
                
                logger.info(f"✅ Threat data logged to blockchain: {tx_hash}")
                return tx_hash
            
            return None
            
        except Exception as e:
            logger.error(f"❌ Failed to log threat data to blockchain: {e}")
            return None
    
    async def verify_data_integrity(self, data_hash: str) -> Dict[str, Any]:
        """
        Verify data integrity using blockchain records
        Args:
            data_hash: Hash to verify
        Returns:
            Verification result with details
        """
        try:
            logger.info(f"🔍 Verifying data integrity: {data_hash[:16]}...")
            
            # Query smart contract for verification
            function_name = "verifyDataIntegrity"
            params = [data_hash]
            
            # Read from smart contract
            result = await self._read_contract_function(function_name, params)
            
            if result:
                verification_result = {
                    'verified': result.get('verified', False),
                    'timestamp': result.get('timestamp'),
                    'source': result.get('source'),
                    'block_number': result.get('blockNumber'),
                    'transaction_hash': result.get('transactionHash')
                }
                
                logger.info(f"✅ Data verification completed: {verification_result['verified']}")
                return verification_result
            
            return {'verified': False, 'error': 'No blockchain record found'}
            
        except Exception as e:
            logger.error(f"❌ Data verification failed: {e}")
            return {'verified': False, 'error': str(e)}
    
    async def get_audit_trail(
        self, 
        start_date: str, 
        end_date: str, 
        data_type: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Get audit trail of blockchain transactions
        Args:
            start_date: Start date (ISO format)
            end_date: End date (ISO format)
            data_type: Optional filter by data type
        Returns:
            List of blockchain transactions
        """
        try:
            logger.info(f"📋 Retrieving audit trail: {start_date} to {end_date}")
            
            # Get events from smart contract
            events = await self._get_contract_events(start_date, end_date)
            
            # Filter by data type if specified
            if data_type:
                events = [e for e in events if e.get('data_type') == data_type]
            
            # Enhance with database records
            enhanced_events = []
            for event in events:
                db_record = await self._get_blockchain_record(event.get('transactionHash'))
                if db_record:
                    event.update(db_record)
                enhanced_events.append(event)
            
            logger.info(f"✅ Retrieved {len(enhanced_events)} audit trail records")
            return enhanced_events
            
        except Exception as e:
            logger.error(f"❌ Failed to retrieve audit trail: {e}")
            return []
    
    async def _execute_contract_function(
        self, 
        function_name: str, 
        params: List
    ) -> Optional[str]:
        """Execute a smart contract function via Starton"""
        try:
            url = f"{self.base_url}/smart-contract/{self.network}/{self.contract_address}/call"
            
            payload = {
                "functionName": function_name,
                "params": params,
                "signerWallet": await self._get_signer_wallet(),
                "speed": "average"
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    url, 
                    headers=self.headers, 
                    json=payload,
                    timeout=aiohttp.ClientTimeout(total=30)
                ) as response:
                    
                    if response.status == 201:
                        result = await response.json()
                        return result.get('transactionHash')
                    else:
                        error_text = await response.text()
                        logger.error(f"Starton API error {response.status}: {error_text}")
                        return None
                        
        except Exception as e:
            logger.error(f"Contract execution error: {e}")
            return None
    
    async def _read_contract_function(
        self, 
        function_name: str, 
        params: List
    ) -> Optional[Dict]:
        """Read from a smart contract function via Starton"""
        try:
            url = f"{self.base_url}/smart-contract/{self.network}/{self.contract_address}/read"
            
            payload = {
                "functionName": function_name,
                "params": params
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    url, 
                    headers=self.headers, 
                    json=payload,
                    timeout=aiohttp.ClientTimeout(total=15)
                ) as response:
                    
                    if response.status == 200:
                        return await response.json()
                    else:
                        logger.error(f"Contract read error {response.status}")
                        return None
                        
        except Exception as e:
            logger.error(f"Contract read error: {e}")
            return None
    
    async def _get_contract_events(
        self, 
        start_date: str, 
        end_date: str
    ) -> List[Dict[str, Any]]:
        """Get events from smart contract"""
        try:
            url = f"{self.base_url}/smart-contract/{self.network}/{self.contract_address}/events"
            
            params = {
                'fromBlock': 'earliest',
                'toBlock': 'latest',
                'limit': 1000
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    url, 
                    headers=self.headers, 
                    params=params
                ) as response:
                    
                    if response.status == 200:
                        data = await response.json()
                        return data.get('items', [])
                    else:
                        logger.error(f"Events fetch error {response.status}")
                        return []
                        
        except Exception as e:
            logger.error(f"Events fetch error: {e}")
            return []
    
    async def _get_signer_wallet(self) -> str:
        """Get or create signer wallet address for transactions"""
        try:
            # In production, this would be a secure wallet management system
            # For demo, we'll use a test wallet
            return "0x742d35Cc6634C0532925a3b8D395DBFAF4fB8fE4"  # Test wallet
            
        except Exception as e:
            logger.error(f"Error getting signer wallet: {e}")
            return ""
    
    async def _store_blockchain_record(
        self, 
        tx_hash: str, 
        data_hash: str, 
        timestamp: str, 
        source: str, 
        data_type: str,
        reference_id: Optional[str] = None
    ):
        """Store blockchain transaction record in database"""
        try:
            record = {
                'transaction_hash': tx_hash,
                'data_hash': data_hash,
                'timestamp': timestamp,
                'source': source,
                'data_type': data_type,
                'reference_id': reference_id,
                'network': self.network,
                'contract_address': self.contract_address,
                'created_at': datetime.utcnow().isoformat()
            }
            
            await self.supabase.table('blockchain_transactions').insert(record).execute()
            logger.info(f"📝 Blockchain record stored: {tx_hash}")
            
        except Exception as e:
            logger.error(f"Failed to store blockchain record: {e}")
    
    async def _get_blockchain_record(self, tx_hash: str) -> Optional[Dict]:
        """Get blockchain transaction record from database"""
        try:
            result = await self.supabase.table('blockchain_transactions')\
                .select('*')\
                .eq('transaction_hash', tx_hash)\
                .execute()
            
            if result.data:
                return result.data[0]
            return None
            
        except Exception as e:
            logger.error(f"Failed to get blockchain record: {e}")
            return None
    
    def _create_data_hash(self, data: Any) -> str:
        """Create SHA-256 hash of data"""
        try:
            if isinstance(data, dict):
                data_str = json.dumps(data, sort_keys=True, separators=(',', ':'))
            else:
                data_str = str(data)
            
            return hashlib.sha256(data_str.encode()).hexdigest()
            
        except Exception as e:
            logger.error(f"Error creating data hash: {e}")
            return hashlib.sha256(str(data).encode()).hexdigest()
    
    async def get_blockchain_statistics(self) -> Dict[str, Any]:
        """Get blockchain usage statistics"""
        try:
            result = await self.supabase.table('blockchain_transactions')\
                .select('*', count='exact')\
                .execute()
            
            total_transactions = result.count
            
            # Get counts by data type
            type_counts = await self.supabase.rpc(
                'get_blockchain_stats_by_type'
            ).execute()
            
            return {
                'total_transactions': total_transactions,
                'network': self.network,
                'contract_address': self.contract_address,
                'transactions_by_type': type_counts.data if type_counts.data else {},
                'last_updated': datetime.utcnow().isoformat()
            }
            
        except Exception as e:
            logger.error(f"Error getting blockchain statistics: {e}")
            return {
                'total_transactions': 0,
                'error': str(e)
            }

File: frontend/bubble_integration/js/blockchain.js
/**
 * Ocean Sentinel - Blockchain Integration
 * Web3 and Starton blockchain verification client
 */

class BlockchainClient {
    constructor() {
        this.config = {
            startonApiKey: window.STARTON_API_KEY,
            contractAddress: window.CONTRACT_ADDRESS,
            network: window.BLOCKCHAIN_NETWORK || 'mumbai',
            startonBaseUrl: 'https://api.starton.io/v3'
        };
        
        this.cache = new Map();
        this.verificationQueue = [];
        this.isProcessingQueue = false;
        
        this.init();
    }
    
    async init() {
        console.log('⛓️ Initializing blockchain client...');
        
        // Verify configuration
        if (!this.config.startonApiKey || !this.config.contractAddress) {
            console.warn('⚠️ Blockchain configuration incomplete');
            return;
        }
        
        try {
            // Test connection to Starton API
            await this.testConnection();
            
            // Start processing verification queue
            this.startQueueProcessor();
            
            console.log('✅ Blockchain client initialized');
            
        } catch (error) {
            console.error('❌ Blockchain client initialization failed:', error);
        }
    }
    
    async testConnection() {
        const response = await fetch(`${this.config.startonBaseUrl}/smart-contract`, {
            headers: {
                'x-api-key': this.config.startonApiKey,
                'Content-Type': 'application/json'
            }
        });
        
        if (!response.ok) {
            throw new Error(`Starton API connection failed: ${response.status}`);
        }
        
        console.log('✅ Starton API connection verified');
    }
    
    async verifyDataIntegrity(dataHash, showUI = true) {
        try {
            if (showUI) {
                this.showVerificationSpinner(dataHash);
            }
            
            // Check cache first
            const cacheKey = `verify_${dataHash}`;
            if (this.cache.has(cacheKey)) {
                const result = this.cache.get(cacheKey);
                if (showUI) {
                    this.showVerificationResult(dataHash, result);
                }
                return result;
            }
            
            // Call smart contract verification function
            const result = await this.callSmartContractRead(
                'verifyDataIntegrity',
                [dataHash]
            );
            
            const verificationResult = {
                verified: result?.verified || false,
                timestamp: result?.timestamp || null,
                blockNumber: result?.blockNumber || null,
                transactionHash: result?.transactionHash || null,
                source: result?.source || null
            };
            
            // Cache result
            this.cache.set(cacheKey, verificationResult);
            
            if (showUI) {
                this.showVerificationResult(dataHash, verificationResult);
            }
            
            return verificationResult;
            
        } catch (error) {
            console.error('Blockchain verification failed:', error);
            
            if (showUI) {
                this.showVerificationError(dataHash, error);
            }
            
            return { verified: false, error: error.message };
        }
    }
    
    async getAuditTrail(startDate, endDate, dataType = null) {
        try {
            const events = await this.getContractEvents('DataLogged', {
                fromBlock: 'earliest',
                toBlock: 'latest',
                limit: 1000
            });
            
            // Filter events by date range and type
            let filteredEvents = events.filter(event => {
                const eventDate = new Date(event.timestamp);
                return eventDate >= new Date(startDate) && eventDate <= new Date(endDate);
            });
            
            if (dataType) {
                filteredEvents = filteredEvents.filter(event => 
                    event.args?.dataType === dataType
                );
            }
            
            return filteredEvents.map(event => ({
                transactionHash: event.transactionHash,
                blockNumber: event.blockNumber,
                timestamp: event.timestamp,
                dataHash: event.args?.dataHash,
                source: event.args?.source,
                dataType: event.args?.dataType
            }));
            
        } catch (error) {
            console.error('Failed to get audit trail:', error);
            return [];
        }
    }
    
    async callSmartContractRead(functionName, params = []) {
        const url = `${this.config.startonBaseUrl}/smart-contract/${this.config.network}/${this.config.contractAddress}/read`;
        
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'x-api-key': this.config.startonApiKey,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                functionName,
                params
            })
        });
        
        if (!response.ok) {
            throw new Error(`Smart contract read failed: ${response.status}`);
        }
        
        return await response.json();
    }
    
    async getContractEvents(eventName, options = {}) {
        const url = `${this.config.startonBaseUrl}/smart-contract/${this.config.network}/${this.config.contractAddress}/events`;
        
        const queryParams = new URLSearchParams({
            eventName,
            ...options
        });
        
        const response = await fetch(`${url}?${queryParams}`, {
            headers: {
                'x-api-key': this.config.startonApiKey,
                'Content-Type': 'application/json'
            }
        });
        
        if (!response.ok) {
            throw new Error(`Get contract events failed: ${response.status}`);
        }
        
        const data = await response.json();
        return data.items || [];
    }
    
    async getBulkVerificationStatus(dataHashes) {
        const results = [];
        
        for (const hash of dataHashes) {
            try {
                const result = await this.verifyDataIntegrity(hash, false);
                results.push({ hash, ...result });
            } catch (error) {
                results.push({ hash, verified: false, error: error.message });
            }
        }
        
        return results;
    }
    
    queueVerification(dataHash) {
        if (!this.verificationQueue.includes(dataHash)) {
            this.verificationQueue.push(dataHash);
        }
    }
    
    startQueueProcessor() {
        if (this.isProcessingQueue) return;
        
        this.isProcessingQueue = true;
        
        const processQueue = async () => {
            while (this.verificationQueue.length > 0) {
                const dataHash = this.verificationQueue.shift();
                try {
                    await this.verifyDataIntegrity(dataHash, false);
                    // Small delay to avoid rate limiting
                    await new Promise(resolve => setTimeout(resolve, 100));
                } catch (error) {
                    console.error(`Queue verification failed for ${dataHash}:`, error);
                }
            }
            
            // Check again after 30 seconds
            setTimeout(() => {
                if (this.verificationQueue.length > 0) {
                    processQueue();
                }
            }, 30000);
        };
        
        processQueue();
    }
    
    showVerificationSpinner(dataHash) {
        const shortHash = dataHash.substring(0, 8) + '...';
        
        // Create or update verification status element
        let statusElement = document.getElementById(`verification-${dataHash}`);
        if (!statusElement) {
            statusElement = document.createElement('div');
            statusElement.id = `verification-${dataHash}`;
            statusElement.className = 'blockchain-verification-status';
        }
        
        statusElement.innerHTML = `
            <div class="verification-item verifying">
                <span class="verification-icon">⏳</span>
                <span class="verification-text">Verifying ${shortHash}</span>
                <div class="verification-spinner"></div>
            </div>
        `;
        
        // Append to verification container or create one
        let container = document.getElementById('blockchain-verifications');
        if (!container) {
            container = document.createElement('div');
            container.id = 'blockchain-verifications';
            container.className = 'blockchain-verifications-container';
            document.body.appendChild(container);
        }
        
        container.appendChild(statusElement);
    }
    
    showVerificationResult(dataHash, result) {
        const shortHash = dataHash.substring(0, 8) + '...';
        const statusElement = document.getElementById(`verification-${dataHash}`);
        
        if (!statusElement) return;
        
        const isVerified = result.verified;
        const iconClass = isVerified ? 'verified' : 'unverified';
        const icon = isVerified ? '✅' : '❌';
        const status = isVerified ? 'Verified' : 'Not Found';
        
        statusElement.innerHTML = `
            <div class="verification-item ${iconClass}">
                <span class="verification-icon">${icon}</span>
                <span class="verification-text">${status}: ${shortHash}</span>
                ${isVerified ? `
                    <div class="verification-details">
                        <small>Block: ${result.blockNumber || 'N/A'}</small>
                        <small>Time: ${result.timestamp ? new Date(result.timestamp).toLocaleString() : 'N/A'}</small>
                    </div>
                ` : ''}
            </div>
        `;
        
        // Auto-remove after 5 seconds if verified, 10 seconds if not
        setTimeout(() => {
            if (statusElement.parentElement) {
                statusElement.remove();
            }
        }, isVerified ? 5000 : 10000);
    }
    
    showVerificationError(dataHash, error) {
        const shortHash = dataHash.substring(0, 8) + '...';
        const statusElement = document.getElementById(`verification-${dataHash}`);
        
        if (!statusElement) return;
        
        statusElement.innerHTML = `
            <div class="verification-item error">
                <span class="verification-icon">⚠️</span>
                <span class="verification-text">Error: ${shortHash}</span>
                <div class="verification-details">
                    <small>${error.message || 'Verification failed'}</small>
                </div>
            </div>
        `;
        
        // Auto-remove after 10 seconds
        setTimeout(() => {
            if (statusElement.parentElement) {
                statusElement.remove();
            }
        }, 10000);
    }
    
    async generateBlockchainReport(startDate, endDate) {
        try {
            const auditTrail = await this.getAuditTrail(startDate, endDate);
            
            const report = {
                period: { startDate, endDate },
                totalTransactions: auditTrail.length,
                dataTypes: {},
                sources: {},
                dailyStats: {},
                integrityScore: 100 // All blockchain data is verified by definition
            };
            
            // Analyze audit trail
            auditTrail.forEach(event => {
                // Count by data type
                const dataType = event.dataType || 'unknown';
                report.dataTypes[dataType] = (report.dataTypes[dataType] || 0) + 1;
                
                // Count by source
                const source = event.source || 'unknown';
                report.sources[source] = (report.sources[source] || 0) + 1;
                
                // Daily statistics
                const date = new Date(event.timestamp).toDateString();
                report.dailyStats[date] = (report.dailyStats[date] || 0) + 1;
            });
            
            return report;
            
        } catch (error) {
            console.error('Failed to generate blockchain report:', error);
            return null;
        }
    }
    
    createBlockchainWidget(containerId) {
        const container = document.getElementById(containerId);
        if (!container) return;
        
        const widget = document.createElement('div');
        widget.className = 'blockchain-widget';
        widget.innerHTML = `
            <div class="widget-header">
                <h3>🔗 Blockchain Integrity</h3>
                <span class="network-badge">${this.config.network}</span>
            </div>
            <div class="widget-content">
                <div class="integrity-status">
                    <div class="status-item">
                        <span class="status-label">Network:</span>
                        <span class="status-value network-status">Connected</span>
                    </div>
                    <div class="status-item">
                        <span class="status-label">Contract:</span>
                        <span class="status-value contract-address">${this.config.contractAddress.substring(0, 10)}...</span>
                    </div>
                    <div class="status-item">
                        <span class="status-label">Verified Records:</span>
                        <span class="status-value verified-count">Loading...</span>
                    </div>
                </div>
                <div class="widget-actions">
                    <button onclick="blockchainClient.showAuditTrail()">View Audit Trail</button>
                    <button onclick="blockchainClient.verifyCurrentData()">Verify Data</button>
                </div>
            </div>
        `;
        
        container.appendChild(widget);
        
        // Update verified count
        this.updateVerifiedCount();
    }
    
    async updateVerifiedCount() {
        try {
            const events = await this.getContractEvents('DataLogged', {
                fromBlock: 'latest-100',
                limit: 100
            });
            
            const countElement = document.querySelector('.verified-count');
            if (countElement) {
                countElement.textContent = events.length.toString();
            }
            
        } catch (error) {
            console.error('Failed to update verified count:', error);
        }
    }
    
    async showAuditTrail() {
        const endDate = new Date();
        const startDate = new Date(endDate.getTime() - 7 * 24 * 60 * 60 * 1000); // 7 days ago
        
        const auditTrail = await this.getAuditTrail(startDate, endDate);
        
        // Create modal with audit trail
        const modal = document.createElement('div');
        modal.className = 'blockchain-modal';
        modal.innerHTML = `
            <div class="modal-content">
                <div class="modal-header">
                    <h2>Blockchain Audit Trail</h2>
                    <button class="close-modal" onclick="this.parentElement.parentElement.parentElement.remove()">×</button>
                </div>
                <div class="modal-body">
                    <div class="audit-summary">
                        <p><strong>Period:</strong> ${startDate.toDateString()} - ${endDate.toDateString()}</p>
                        <p><strong>Total Transactions:</strong> ${auditTrail.length}</p>
                    </div>
                    <div class="audit-table">
                        <table>
                            <thead>
                                <tr>
                                    <th>Time</th>
                                    <th>Data Hash</th>
                                    <th>Source</th>
                                    <th>Transaction</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${auditTrail.slice(0, 50).map(event => `
                                    <tr>
                                        <td>${new Date(event.timestamp).toLocaleString()}</td>
                                        <td>${event.dataHash?.substring(0, 10)}...</td>
                                        <td>${event.source}</td>
                                        <td>
                                            <a href="https://mumbai.polygonscan.com/tx/${event.transactionHash}" 
                                               target="_blank">${event.transactionHash?.substring(0, 10)}...</a>
                                        </td>
                                    </tr>
                                `).join('')}
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        `;
        
        document.body.appendChild(modal);
    }
    
    async verifyCurrentData() {
        if (window.dashboard?.state?.environmentalData) {
            const dataHash = window.dashboard.state.environmentalData.data_hash;
            if (dataHash) {
                await this.verifyDataIntegrity(dataHash, true);
            }
        }
    }
    
    cleanup() {
        console.log('🧹 Cleaning up blockchain client...');
        this.cache.clear();
        this.verificationQueue = [];
        this.isProcessingQueue = false;
    }
}

// Global blockchain client instance
window.blockchainClient = new BlockchainClient();

// Cleanup on page unload
window.addEventListener('beforeunload', () => {
    if (window.blockchainClient) {
        window.blockchainClient.cleanup();
    }
});
